src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -117,19 +146,240 @@ void optimizeJoin(QueryPlan::Node & node, QueryPlan::Nodes &)
+{
+    bool keep_logical = optimization_settings.keep_logical_steps;
+    auto * join_step = typeid_cast<JoinStepLogical *>(node.step.get());
+    if (!join_step)
+        return false;
+    if (node.children.size() != 2)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "JoinStepLogical should have exactly 2 children, but has {}", node.children.size());
+
+    JoinActionRef left_filter(nullptr);
+    JoinActionRef right_filter(nullptr);
+    JoinActionRef post_filter(nullptr);
+    auto join_ptr = join_step->convertToPhysical(left_filter, right_filter, post_filter, keep_logical);
+
+    if (join_ptr->isFilled())
+    {
+        node.children.pop_back();
+    }
+
+    if (keep_logical)
+        return true;
+
+    Header output_header = join_step->getOutputHeader();
+
+    auto & join_expression_actions = join_step->getExpressionActions();
+
+    QueryPlan::Node * new_left_node = makeExpressionNodeOnTopOf(node.children.at(0), std::move(join_expression_actions.left_pre_join_actions), left_filter.column_name, nodes);
+    QueryPlan::Node * new_right_node = nullptr;
+    if (node.children.size() >= 2)
+        new_right_node = makeExpressionNodeOnTopOf(node.children.at(1), std::move(join_expression_actions.right_pre_join_actions), right_filter.column_name, nodes);
+
+    if (join_step->areInputsSwapped() && new_right_node)
+        std::swap(new_left_node, new_right_node);
+
+    const auto & settings = join_step->getContext()->getSettingsRef();
+
+    auto & new_join_node = nodes.emplace_back();
+
+    if (!join_ptr->isFilled())
+    {
+        chassert(new_right_node);
+        if (const auto * fsmjoin = dynamic_cast<const FullSortingMergeJoin *>(join_ptr.get()))
+            addSortingForMergeJoin(fsmjoin, new_left_node, new_right_node, nodes,
+                join_step->getSortingSettings(), join_step->getJoinSettings(), join_step->getJoinInfo());
+
+        auto required_output_from_join = join_expression_actions.post_join_actions.getRequiredColumnsNames();
+        new_join_node.step = std::make_unique<JoinStep>(
+            new_left_node->step->getOutputHeader(),
+            new_right_node->step->getOutputHeader(),
+            join_ptr,
+            settings[Setting::max_block_size],
+            settings[Setting::min_joined_block_size_bytes],
+            settings[Setting::max_threads],
+            NameSet(required_output_from_join.begin(), required_output_from_join.end()),
+            false /*optimize_read_in_order*/,
+            true /*use_new_analyzer*/);
+        new_join_node.children = {new_left_node, new_right_node};
+    }
+    else
+    {
+        new_join_node.step = std::make_unique<FilledJoinStep>(
+            new_left_node->step->getOutputHeader(),
+            join_ptr,
+            settings[Setting::max_block_size]);
+        new_join_node.children = {new_left_node};
+    }
+
+    {
+        WriteBufferFromOwnString buffer;
```

Comment:
Why do we perform this `describeActions` into temporary buffer which is not used after that?